Update guide.md (#837)
swcarpentry/python-novice-inflammation: _extras/guide.md

## Overall

This lesson is written as an introduction to Python, 
and our goal is to explain to the students that all computer software 
no matter what langauge they are written in is composed of concepts like 
variables, conditionals, looping and functions.
We make a special effort to introduce the functions, the single most important idea in programming
that allows reusability and efficient logical organization of the software.

> We show how to solve problems by building functions,

Reviewer: I wasn't fond of the original language because it is so strongly worded with saying "the single most important idea in programming". I like that you add in a reason for why functions are important. However, I think the new wording takes some of the focus off of problem solving and puts it more on concepts of programming, where I think we want to highlight that we are teaching learners to solve problems with code, not about the code itself. I suggest tweaking the language to put the focus back on problem solving while retaining your explanation of why functions are important. What do you (and anyone else who cares to share) think about:
```
This lesson is written as an introduction to Python,
but its real purpose is to introduce an important idea in programming:
how to solve problems by building functions,
units of code that allow reusability and efficient logical organization of software.
In order to teach that,
we must teach people a little about
programming concepts such as variables, conditionals, loops,
and the mechanics of manipulating data with lists and file I/O
so that their functions can do things they actually care about.
```

Reviewer 2: I like your suggestions. Maybe "file I/O" can be a little intimidating?
How about the following:

```
This lesson is written as an introduction to Python,
but its wider purpose is to introduce an important idea in programming:
how to solve problems by building functions, units of code that allow
reusability and efficient logical organization of software.

So that our functions can do things we actually care about,
we cover programming concepts such as variables, conditionals,
loops, and the mechanics of manipulating data with lists,
and of reading from files and writing to files.
```

Author: I would agree with both points, to make a special emphasis on functions, and to perhaps not to intimidate the potentially non-technical crowds with terms like the file I/O operations. I like @slel's version, since it incorporates both improvements. I should note though, that the training did not have the feeling like there was a particular effort on the introduction and practice of functions. If that's the intent, perhaps the curriculum might need updating sometime as well.

Reviewer: I agree with @slel (thanks for commenting!) about replacing the term "file I/O", though I don't think we really spend time on writing to files. Maybe the second sentence could be:
```
So that our functions can do things we actually care about,
we cover programming concepts such as variables, conditionals,
loops, lists, and the mechanics of manipulating data read from files.
```

Author: Very good, for the sake of simplicity I suggest: 
```suggestion
So that our functions can do things we actually care about,
we cover programming concepts such as variables, conditionals,
loops, lists, and reading data from files.
```

## Overall

> This lesson is written as an introduction to Python,

Reviewer: I don't think a `.` makes sense there @maxim-belkin as the sentence continues after the punctuation.

> This lesson is written as an introduction to Python,

Reviewer 3: This proposed change is related to the one below: https://github.com/swcarpentry/python-novice-inflammation/pull/837#discussion_r450181162

> This lesson is written as an introduction to Python,

Reviewer 3: Resolving in favor of https://github.com/swcarpentry/python-novice-inflammation/pull/837#discussion_r450963457

but its wider purpose is to introduce an important idea in programming:
how to solve problems by building functions, units of code that allow

> reusability and efficient logical organization of software.

Reviewer: I don't really think the sentence needs to change further, but if you are wanting to break it up and change it, I prefer "However" over "Still" and a `,` instead of "and" between "efficient" and "logical". However, I am fine with Maxim's suggestion as-if if that is what you all want. @ndilekli what please commit as you see fit!

> reusability and efficient logical organization of software.

Reviewer 3: I'm OK if we don't break it into two but I think we need to add a comma or an `and` between "efficient" and "logical" and maybe change comma dashes before "the units". The reason I proposed to break this sentence up is that it seems to have too many "stops" / "breaks":
1. `, but`
2. `programming:`
3. `functions,`
4. `efficient [and/,] logical`

... and breaking up the sentence into two had minimal effect on the changes in this PR. Other way is to remove "to introduce an important idea in programming:" so that it reads

> but its wider purpose is to show how to solve problems by building functions...

but this is open up for a debate and, again, we can always leave it as-is here and address it in a future PR.

> reusability and efficient logical organization of software.

Reviewer 3: Resolving in favor of https://github.com/swcarpentry/python-novice-inflammation/pull/837#discussion_r450963457

So that our functions can do things we actually care about,

> we cover programming concepts such as variables, conditionals,

Reviewer 3: "So that our functions can do things we actually care about,"  would've been a great intro to a paragraph describing how to write better functions when we already know how to write simple ones. I think we're one or two steps away from there. Would any of the following fit better the prose of this section?:

> To show learners how to write functions that do things they care about, we cover basic programming concepts such as variables, ...

or
> To teach how write functions that do exactly what we care about, we cover basic programming concepts such as variables, ...

Reviewer: In the context of the Instructor Notes and this being just after a statement about what functions are for, I don't really think the prose options above will make for much more of a better fit for the section. However, I prefer your first option to the second (because I don't favor the use of "exactly") if others feel the wording should be changed.

Reviewer 3: > (because I don't favor the use of "exactly")

Yeah, I think I agree with that.

Anyways, my point was that we cover programming concepts that we cover not to make our functions do things we care about but to teach/learn how to program in Python. We also learn about functions (which should be added to the list of things we learn). So, here is another possible justification:

> To teach how to write Python programs that perform calculations or carry out analysis that we need, we cover most important programming concepts such as variables, conditional statements, loops, lists, functions. We also show how to read data from files and provide input on the command line.

Reviewer: Perhaps you are speaking to the point @ndilekli brought up earlier:

> the training did not have the feeling like there was a particular effort on the introduction and practice of functions.

In this case, perhaps we should take the emphasis off of functions in the previous sentence as well and say something along the lines of:

> This lesson is written as an introduction to Python,
but, its broader purpose is to introduce how to solve problems by developing programs.
By writing lines of instructions to carry out tasks, we automate our work and
enable reproducible results.
To teach how to write Python programs that perform calculations or carry out analysis that we need,
we cover most important programming concepts such as variables, conditional statements, loops,
lists, and functions.
We also show how to read data from files and provide input on the command line.

Reviewer 3: Yes, I like this paragraph a lot more: it's easy to read and makes the right emphasis (I'm not sure about "lines of instructions" in the second sentence -- perhaps we need to clearly state that programs are lines of code/instructions -- but that's a minor detail).

*   It's free.
*   It has a lot of scientific libraries, and more are constantly being added.
*   It has a large scientific user community.
*   It's easier for novices to learn than most of the mature alternatives, 

> especially due to its syntax that reads like a human language.

Reviewer: ```suggestion
partly due to its syntax that reads more like a human language.
```

Reviewer: Hi Naci, that's great you got the changes committed to the branch for the pull request! Regarding this question:

> So I guess whenever a conversation is resolved with some suggestions, that in itself becomes a commit, is that correct?

Clicking the `Resolve conversation` button on a discussion point does not commit anything to your files; it is a mechanism of the GitHub interface to say, "ok, we've made a decision on this and no longer need to discuss it further."

When the maintainers do a review of your pull request, we tend to make comments that suggest ideas for editing your submission, and GitHub allows us to put the comment in-line with the code, so we can all see what we are referencing. Generally this does not automatically update the code, it is giving you our opinion about the code change, and we want you to think about if you want to do this then manually change the file either in the GitHub interface (as you did), or at the command-line using `git`.

It is also possible to make a "suggestion" when doing review of the pull request, where we are actually asking if you want to commit the exact change we are proposing. That is what I have done above with the "partly due to its syntax..." comment, and you should be able to click on the `Commit suggestion` dropdown to commit the changes if you agree with them. In this case of using the `Commit suggestion` button, you do not need to manually edit a file, GitHub will automatically make the commit for you. I believe you used this feature earlier on this pull request. If we did not click on `Commit suggestion` the change would not get automatically committed. If you would like to included the "Suggested change" I have added above this, please try committing the suggestion with the interface above to see how that works. :)

Does that clarify these different aspects of the pull request process?

Author: Lauren, thank you very much, I think I have a much better understanding of the process now. I think I have a relatively steep learning curve since I never worked on SW development in teams, but I can say it sure is a very powerful process based on all these elaborate schemes to put all those inputs and interactions into the development.

> partly due to its syntax that reads more like a human language.

Reviewer 3: I agree that we need to explain why [we think] Python is easier to learn [than other programming languages]. However, these bullet points state the reasons why [we think] Python is better suited for teaching basic Programming skills. So, I think it might be better to not explain ourselves right here but move the explanation to the next paragraph. It could start with something like:

> That last point above is important. Code written in Python reads more like a human language. Originally, Software Carpentry was teaching programming using Perl. When we switched to Python, we found..."

What do you guys think?

Reviewer: I think it is fine as-is. I think it will be easier to miss the point if it is moved to a separate paragraph. Also, the lead-in to the list starts with "Explain", so I think it is okay to explain in the list.

Reviewer 3: > I think it will be easier to miss the point if it is moved to a separate paragraph.

I'm not suggesting to move the entire point to a separate paragraph -- only its justification / explanation. Keeping only the main points in the list should make them easier to remember. Those who don't read the next paragraph will miss the explanation(s) but not the point(s). 

Speaking of learning... I believe it's the combination of syntax and semantics that makes Python easy to learn: syntax is related to writing proper Python code, and semantics -- to making sense of Python code.

> Also, the lead-in to the list starts with "Explain"

To me this serves as a justification to write bullet points as succinct statements rather than sentences. Explanations that we can provide in the list can only be of limited length and I think we shouldn't limit our explanations by the choice of where we provide them.

Do you think the following would convey the message of the last bullet point?

> Its easy-to-learn syntax and semantics make it ideal for teaching basic programming skills.

Reviewer: Taking out from the bullet point the part about Perl as a  justification that Python is easier, now we would be stating Python is "easy-to-learn" which I don't like. What makes for Python being "easy-to-learn syntax and semantics"? It may not be considered easy to someone, especially someone who has no reference to other programming syntax. How are the semantics of Python different than other languages?

Reviewer 3: > Taking out from the bullet point the part about Perl as a justification that Python is easier, now we would be stating Python is "easy-to-learn" which I don't like. 

I guess my last suggestion was not the best and it confused you. Let me try to better explain my concern. 
What we have now in this PR can be represented as:

> we do X because:
> A
> B
> C, partially due to D (E)

This last bullet point seems to be  overloaded with information a little bit, so I propose to say:

> we do X because
> A
> B
> C
>
> C is partially due to D which, in turn, is supported by E.

(A, B, C) together explain why we _teach_ programming using Python.
(C) is the point about Python being easier to _learn_ than other mature alternatives.
I think (C) is important enough to justify spending a paragraph on it.

> How are the semantics of Python different than other languages?

In Python `"1" + 2` is valid syntactically but not semantically.
In Perl, the same expression is valid syntactically and semantically (result is 3).

In Python `"1" + "2"` is valid syntactically and semantically.
In Perl, the identical expression (`"1" + "2"`) is valid syntactically but not semantically (result is 0) and syntax of a semantically valid equivalent is `"1" . "2"`.

So, it's always a combination of syntax and semantics that makes a language intuitive or... not-so-much. I think that the closer the syntax and semantics of a programming language to those of (human language + mathematics), the more intuitive it is. 

> What makes for Python being "easy-to-learn syntax and semantics"? It may not be considered easy to someone, especially someone who has no reference to other programming syntax.

Agreed, this was not the best wording -- we don't have to use it and we can keep "It's easier to learn than most of the mature alternatives". I'm mostly concerned with overloading the last point as I tried to explain in the beginning of this comment.

Reviewer: Perhaps you could do a pull request to submit your changes there @maxim-belkin if you are wanting to reword/reorganize beyond what is presented in this pull request . I think we may have lost @ndilekli :/.

Author: Hello, sorry that I was out of the loop for a while. I could do another commit, whenever you would like me to. Regarding the Python syntax, I could comment that as someone starting out with Basic, then continuing with Pascal and C, the syntax of Python is indeed much more like reading a text in English. Saying "reads like English" might be an overstatement, but I have seen quite a lot of remarks on that aspect. We might perhaps instead say it is more readable relative to many other programming languages like C and Java. What do you think?

Author: Hello, sorry that I was out of the loop for a while. I could do another commit, whenever you would like me to. Regarding the Python syntax, I could comment that as someone starting out with Basic, then continuing with Pascal and C, the syntax of Python is indeed much more like reading a text in English. Saying "reads like English" might be an overstatement, but I have seen quite a lot of remarks on that aspect. We might perhaps instead say it is more readable relative to many other programming languages like C and Java. What do you think?

Reviewer 3: > Hello, sorry that I was out of the loop for a while. 

No worries!

> I could do another commit, whenever you would like me to.

Feel free to update the PR whenever you like if you agree with the comments / suggestions. To sum up:

1. My concern is that we are overloading the last item in the list that explains why we use Python for teaching with information why Python is easier to learn than other (mature) programming languages. Those are related but different topics. So, I propose to move the explanation to the paragraph right after the list. I'm not sure if Lauren (@ldko) agrees with that, so I'd like to clarify this first. If there is no consensus, we might have to leave it as is and discuss this part later in a separate PR or issue.

2. Please have a look at Lauren's suggestion on how to rephrase the first paragraph in introduction: https://github.com/swcarpentry/python-novice-inflammation/pull/837#discussion_r450963457. You can use it "as is" or make modifications you deem important.

>  as someone starting out with Basic, then continuing with Pascal and C

Huh, I took the same path with a quick stop at Visual Basic between Pascal and C (it was interesting to observe how Python adopted and extended Pascal's `:=` assignment operator).

> , the syntax of Python is indeed much more like reading a text in English. Saying "reads like English" might be an overstatement, but I have seen quite a lot of remarks on that aspect. We might perhaps instead say it is more readable relative to many other programming languages like C and Java.

Sure, that'll work too (I think). This way we won't have to go deep into syntax/semantics stuff.

Thank you, Naci!

Reviewer: If you all agree about splitting off some of the last item in the list to a paragraph that follows, that is fine with me. Please go ahead.

*   It's free.
*   It has a lot of scientific libraries, and more are constantly being added.
*   It has a large scientific user community.
*   It's easier for novices to learn than most of the mature alternatives, 

> especially due to its syntax that reads like a human language.

Reviewer: What do you think of toning this down a little to say:
```
*   It's easier for novices to learn than most of the mature alternatives, 
partially due to its syntax reading more like a human language.
```

Reviewer 2: partly?

*   Create a new notebook.
*   Enter code or data in a cell and execute it.
*   Explain the difference between `In[#]` and `Out[#]`.

> *   Show that modifying code in an earlier cell requires re-running any subsequent cells in order for it to reflect the change.

Reviewer: One more thing @ndilekli , we try to keep the lines less than 100 characters in the markdown (see #498 for explanation). Here I am suggesting the change to shorten the line and to change "it" to "them". Thanks again for all your work on this! It's good to know it is teaching you something about collaborating on GitHub.
```suggestion
*   Show that modifying code in an earlier cell requires re-running any subsequent
    cells in order for them to reflect the change.
```

Author: Sure thing!

*   Create a new notebook.
*   Enter code or data in a cell and execute it.
*   Explain the difference between `In[#]` and `Out[#]`.
*   Explain that the cells are independent blocks 
and if we make a change at a top cell we need to run all the subsequent

> cells to make sure the change is reflected across the script.

Reviewer: I think this is a concept worth including, as it often comes up in workshops. I do wonder, though, if teaching it at the very beginning of a lesson might start to overload learners and be confusing to see before people new to programming have seen much code. If people want to include it, I suggest possibly rewriting it more as an operation:
```
* Show that modifying code in an earlier cell requires re-running any subsequent
cell in order for it to reflect the change.
```

Author: Much better wording, thanks! I think you point out to an interesting concern: perhaps this guide can be eventually restructured so that it roughly explains which points correspond to which portions of the curriculum because some particular issues like this one can be confusing at the beginning, but they are the things that better be provided to the students at some point in a curriculum with such scope with fairly complex workflows. To keep things relatively simple, we could flag some sentences in the following way:

```suggestion
* After few episodes show that modifying code in an earlier cell requires re-running any subsequent cells in order for it to reflect the change.
```

Reviewer: I don't think it needs to be a few episodes in, maybe leaving it open-ended is ok, allowing the instructor to teach it when there are adequate cells for the concept. We should be sure to match plural (cells/them) though:
```
* Show that modifying code in an earlier cell requires re-running any
subsequent cells in order for them to reflect the change.
```
